File: 프로그래머스/2/12952. N－Queen/N－Queen.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>

int	queen_ck(int array[], int row, int column)
{
	int	i;

	i = 0;
	while (i < column)
	{
		if (array[i] == row)
			return (0);
		if (array[i] - row == i - column)
			return (0);
		if (array[i] - row == column - i)
			return (0);
		i++;
	}
	return (1);
}

void	recursive(int array[], int *count, int row, int column, int n)
{
	int		i;

	if (column == n)
	{
		*count += 1;
		return ;
	}
	i = row;
	while (i < n)
	{
		if (queen_ck(array, i, column))
		{
			array[column] = i;
			recursive(array, count, 0, column + 1, n);
			if (i + 1 < n)
				recursive(array, count, i + 1, column, n);
			return ;
		}
		i++;
	}
	return ;
}


int solution(int n) {
    int answer = 0;
    int array[n];
    
    recursive(array, &answer, 0, 0, n);
    return answer;
}

File: 프로그래머스/2/12952. N－Queen/N－Queen.py
def queen_ck(array, row, column):
    for i in range(column):
        if array[i] == row:
            return False
        if abs(array[i] - row) == abs(i - column):
            return False
    return True

def recursive(array, count, row, column, n):
    if column == n:
        count[0] += 1
        return

    for i in range(row, n):
        if queen_ck(array, i, column):
            array[column] = i
            recursive(array, count, 0, column + 1, n)
            if i + 1 < n:
                recursive(array, count, i + 1, column, n)
            return

def solution(n):
    count = [0]  # 리스트를 사용하여 참조 가능하게 설정
    array = [-1] * n  # 퀸의 위치를 저장하는 배열
    recursive(array, count, 0, 0, n)
    return count[0]